### ISSUE TITLE
Dynamic objects for update/insert/select

### ISSUE THREAD

**Issue by shelru** [sidorares/node-mysql2#958]
comment

**Comment by shelru** [comment]
https://www.npmjs.com/package/flexqp

Is there anyway to make something like this in mysql2?
Ex: 
```js
req.body.id=1, req.body.params.access = 'abcd' , req.body.params.name='abcd'
execute('update ... set ? where id = ?', [req.body.params, req.body.id]);
//update ... set access = 'abcd', name = 'abcd' where id = 1
```
or 
```js
req.body.id=1, req.body.params.access = 'abcd'
//update ... set access = 'abcd'```

@sidorares 
Было бы здорово передавать лишь нужные поля для обновления. К пример обновить только имя пользователя, либо имя и пароль, либо получить только имя пользователя и его права. Генерировать строки на ходу - не очень привлекательная идея.

**Comment by sidorares** [comment]
no, we don't build or interpolate sql for prepared statements on the client side, it's done by the server and because of this the only syntax supported is what server accepts for prepared statements ( simple types as parameters + same number of placeholders as numbers of parameters ). You can partially work around this by allowing parameter ( first in your example ) to be json string, then pass it to stored procedure, parse JSON and extract data

**Comment by sidorares** [comment]
maybe I misunderstand original question is only about prepared statements and `connection.execute()`

With non-prepared statements `.query()` already allows exactly what you described, see https://github.com/mysqljs/sqlstring readme:

>  - Arrays are turned into list, e.g. ['a', 'b'] turns into 'a', 'b'
> - Nested arrays are turned into grouped lists (for bulk inserts), e.g. [['a', 'b'], ['c', 'd']] turns into ('a', 'b'), ('c', 'd')
> - Objects that have a toSqlString method will have .toSqlString() called and the returned value is used as the raw SQL.
> - Objects are turned into key = 'val' pairs for each enumerable property on the object. If the property's value is a function, it is skipped; if the property's value is an object, toString() is called on it and the returned value is used.

**Comment by shelru** [comment]
Yeah, i get it, but we can interpolate few prepared statements on sside, ex.: ```SELECT ? ? ? WHERE id = ?
SELECT id,name WHERE id = ?, UPDATE users SET id=?, name=? WHERE id = ?,UPDATE users SET id=?, name=?,surename=? WHERE id = ?```.

But ok.

**Comment by shelru** [comment]
@sidorares what is the difference between ? and ?? ?

**Comment by sidorares** [comment]
You can mix client side and server side interpolation at your own risk:

```js
const partialQuery = mysql.format('update ... set ? where id = ?', [yourSetObject]);
conn.execute(partialQuery, [yourId])
```

> @sidorares what is the difference between ? and ?? ?

"Alternatively, you can use ?? characters as placeholders for identifiers" - https://github.com/mysqljs/sqlstring#escaping-query-identifiers

**Comment by shelru** [comment]
Im gonna try, thanks.

**Comment by shelru** [comment]
@sidorares one more question,
 i have bit column
![image](https://user-images.githubusercontent.com/9306996/57073352-ea940c80-6cf1-11e9-9e73-057c83e15f8f.png)
And have prepared statement(execute), when im passing int to query(ex: ? ...,[255]) mysql convert it to bit, how i can know which algoritm is using there? I wanna reproduce it with query, but seem's it's doesnt work.

//Если кратко, можно ли каким-нибудь образом узнать что именно происходит при выполнении запроса, какие преобразования и т.д.

**Comment by shelru** [comment]
But maybe nodejs mysql driver convert int to bit?

**Comment by shelru** [comment]
@sidorares 
```js
const UpdateUserInfoQuery =
    'UPDATE users \
        SET `name`=IFNULL(?,`name`),`surename`=IFNULL(?,`surename`),`fname`=IFNULL(?,`fname`), \
        `login`=IFNULL(?,`login`),`password`=IFNULL(?,`password`),`group`=IFNULL(?,`group`), \
        `type`=IFNULL(?,`type`),`active`=IFNULL(?,`active`),`login_changed`=IFNULL(?,`login_changed`),`type2`=IFNULL(?,`type2`) WHERE `id`=?';
```
maybe add to readme?

**Comment by sidorares** [comment]
> maybe must be added to readme?

what exactly?

**Comment by shelru** [comment]
Like dynamic uppdate's select's and etc.

**Comment by sidorares** [comment]
looks like pretty standard prepared statement to me

**Comment by shelru** [comment]
So, can you show SELECT and INSERT query with dynamic params .execute without .format, at pure sql?

**Comment by sidorares** [comment]
there is no format in your example

**Comment by shelru** [comment]
i know, but is there anyway to get columns dynamic too?

**Comment by sidorares** [comment]
not with execute

all restrictions documented here: https://dev.mysql.com/doc/refman/8.0/en/sql-syntax-prepared-statements.html

**Comment by shelru** [comment]
Seem's like someone did it.
![image](https://user-images.githubusercontent.com/9306996/57165779-27065c00-6e09-11e9-9c73-7f9fdab04e3c.png)

Is there anyway to destroy prepared statement in mysql2?
Like `DEALLOCATE`.

**Comment by shelru** [comment]
`select if(?=1,`name`,null) as `name`, if(?=1,`sname`,null)  as `sname` from users where id = 1;`
Seems i can do something like this
Then send args:
1, 1:
![image](https://user-images.githubusercontent.com/9306996/57166952-ee688180-6e0c-11e9-989d-55f7bf724d5e.png)

0, 1:
![image](https://user-images.githubusercontent.com/9306996/57166912-bfeaa680-6e0c-11e9-9d62-24d29330033e.png)

1, 0:
![image](https://user-images.githubusercontent.com/9306996/57166943-e6104680-6e0c-11e9-8a9a-d90fd72de84e.png)

0, 0:
![image](https://user-images.githubusercontent.com/9306996/57166927-d1cc4980-6e0c-11e9-81f4-06255d7c4f4d.png)

**Comment by shelru** [comment]
![image](https://user-images.githubusercontent.com/9306996/57172142-36979c00-6e2d-11e9-8f6e-b99ffc010dda.png)

**Comment by 2naive** [comment]
`pool.query('INSERT INTO ?? SET ?', [table, object])`

**Comment by sidorares** [comment]
@2naive can you elaborate? I expect your example to work with `.query()`

**Comment by 2naive** [comment]
@sidorares I just left it here as an example for those who will be looking for a shorthand inserting/updating objects. Similar to connection.insert from https://github.com/tshemsedinov/node-mysql-utilities.

So as I see some messages above were deleted and you don't have to use format with execute. That's why one-liner above would be useful for people like me, that find this issue.
